cmd/node-agent/main.go

```diff
@@ -0,0 +1,290 @@
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	if configHasChanged(onDiskConfig, desiredConf) {
+		if err := validateConfig(desiredConf); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err := updateKubeletEnvFile(desiredConf, onDiskConfig); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err := restartKubeletService(); err != nil {
+			glog.Error(err)
+			return
+		}
+	}
+	// Always update annotation at end of sync loop.
+	if err := updateNodeAnnotation(newNode, desiredConf); err != nil {
+		glog.Error(err)
+	}
+}
+
+func configHasChanged(onDiskConfig, desiredConfig map[string]string) bool {
+	if onDiskConfig[kubeletVersionKey] != desiredConfig[kubeletVersionKey] {
+		return true
+	}
+	cl, err := configMapToCommandLineFlags(desiredConfig[kubeletConfigKey])
+	if err != nil {
+		glog.Errorf("error while parsing configMap: %v", err)
+		return false
+	}
+	if onDiskConfig[kubeletFlagsKey] != cl {
+		return true
+	}
+	return false
+}
+
+func validateConfig(desiredConfig map[string]string) error {
+	if _, ok := desiredConfig[kubeletVersionKey]; !ok {
+		return fmt.Errorf("configuration annotation does not contain required key: %s", kubeletVersionKey)
+	}
+	if _, ok := desiredConfig[kubeletConfigKey]; !ok {
+		return fmt.Errorf("configuration annotation does not contain required key: %s", kubeletConfigKey)
+	}
+	return nil
+}
+
+func updateKubeletEnvFile(conf, onDiskConfig map[string]string) error {
+	var buf bytes.Buffer
+	onDiskConfig[kubeletVersionKey] = conf[kubeletVersionKey]
+	commandLineString, err := configMapToCommandLineFlags(conf[kubeletConfigKey])
+	if err != nil {
+		return err
+	}
+	onDiskConfig[kubeletFlagsKey] = commandLineString
+	for k, v := range onDiskConfig {
+		buf.WriteString(fmt.Sprintf("%s=%s\n", k, v))
+	}
+	return writeAndAtomicCopy(buf.Bytes(), kubeletEnvPath)
+}
+
+func parseKubeletEnvFile(path string) (map[string]string, error) {
+	env := make(map[string]string)
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		kv := strings.SplitN(s.Text(), "=", 2)
+		if len(kv) != 2 {
+			glog.Infof("invalid config: %s", s.Text())
+			continue
+		}
+		env[kv[0]] = kv[1]
+	}
+	return env, nil
+}
+
+func updateNodeAnnotation(node *v1.Node, conf map[string]string) error {
+	newConf, err := json.Marshal(conf)
+	if err != nil {
+		return err
+	}
+	node.Annotations[currentConfigAnnotation] = string(newConf)
+	_, err = client.Core().Nodes().Update(node)
+	return err
+}
+
+func restartKubeletService() error {
+	// systemd daemon-reload equivalent
+	err := sysdConn.Reload()
+	if err != nil {
+		return err
+	}
+	ch := make(chan string)
+	_, err = sysdConn.RestartUnit("kubelet.service", "replace", ch)
+	if err != nil {
+		return err
+	}
+	// Wait for unit to restart.
+	result := <-ch
+	if result != "done" {
+		return err
+	}
+	return nil
+}
+
+func configMapToCommandLineFlags(configMapName string) (string, error) {
```

Comment:
Sorry, this is just retrieving from local cache. But it still seems like an overloaded function, where we should be able to have something like `getConfigMap` and `configMapToCommandLine` as isolated functions.